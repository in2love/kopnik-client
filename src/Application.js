import AsyncLock from 'async-lock'
import {AbstractSync, Kopnik, Kopa} from "./models";
import {KopnikApiError, KopnikError} from "./KopnikError";
import once from "./decorators/once";
import SquadAnalyzer from "./SquadAnalyzer";


export default class Application {
    constructor(logger) {
        this.logger = logger.getLogger('application')
        /**
         * Кэш моделей
         * @type {Array}
         */
        // пока не понятно зачем приложению кэш.
        // он вероятно будет сильно торможить приложение из-за реактивности App.$data в которой он находится
        // хотя он же не в App.$data а в App.$options !!
        // уже в дата потому тчо в options не реактивен
        //this.models = AbstractSync.cache

        /**
         * Идентификатор раздела
         * Равен названиею соответствующего компонента
         *
         * @type {string} Map | Profile | Thanks
         */
        this.section = Application.Section.Main
        this.sectionLocker = new AsyncLock

        /**
         * 0 потому что на undefined не срабатывает watcher,
         * а null занят под "пользователь не аутентифицирован"
         */
        this.user = undefined

        this.top20 = []
        this.squadAnalyzer = new SquadAnalyzer()
        this.kopa = new Kopa
        /**
         *
         * @type {Kopnik}
         */
        this.selected = null
        /**
         * Сообщения пользователю
         * @type {String[]}
         */
        this.infos= []
        /**
         * Обишки приложения
         * @type {Error[]}
         */
        this.errors= []
    }

    /**
     * Эксклюзивный доступ к section
     * @param {function} callback
     * @returns {Promise<any>}
     */
    lockSection(callback) {
        return this.sectionLocker.acquire('section', callback)
    }

    /**
     * Устанавливить секцию
     * Если секция не может быть установлена, возвращает иную
     *
     * @param {String} section
     * @returns {Promise<string>} Установленная секция
     */
    async setSection(section) {
        let result
        if (this.section === section) {
            return section
        }
        switch (section) {
            case Application.Section.Profile:
            case Application.Section.Witness:
            case Application.Section.Thanks:
                if (await this.resolveUser()) {
                    if (section===Application.Section.Witness && this.user.status!== Kopnik.Status.CONFIRMED){
                        result= await this.setSection(Application.Section.Main)
                    }
                    else {
                        result = section
                    }
                } else {
                    result = await this.setSection(Application.Section.Main)
                }
                break
            case Application.Section.Main:
                result = section
                break
            default:
                throw new KopnikError('Wrong route', 666)
        }
        // this.logger.info('move', this.section, '->', section)
        return this.section = result
    }

    /**
     * Исследует дружину
     * @param {Kopnik} kopnik
     * @returns {Promise<void>}
     */
    async analyzeSquad(kopnik) {
        await this.squadAnalyzer.analyze(kopnik)
    }

    async loadTop20() {
            this.top20 = await Promise.all([1, 2, 3, 4].map(each => Kopnik.get(each)))
            this.logger.warn('manual set foremans')
            Kopnik.getReference(1).rank = 4
            Kopnik.getReference(2).foreman = Kopnik.getReference(3)
            Kopnik.getReference(2).rank = 1
            Kopnik.getReference(3).foreman = Kopnik.getReference(1)
            Kopnik.getReference(3).rank = 3
            Kopnik.getReference(4).foreman = Kopnik.getReference(3)
            Kopnik.getReference(4).rank = 1

            Kopnik.getReference(1).ten = [Kopnik.getReference(3)]
            Kopnik.getReference(2).ten = []
            Kopnik.getReference(3).ten = [Kopnik.getReference(2), Kopnik.getReference(4)]
            Kopnik.getReference(4).ten = []
    }

    /**
     * Определиться с пользователем. Или он некий копник, или он null, то есть незарегистрирован на сервере
     * @returns {Promise<Kopnik>}
     */
    async resolveUser() {
        if (this.user === undefined) {
            await this.authenticate()
        }
        return this.user
    }

    getSharedState() {
        return {
            SECTION: this.section
        }
    }

    setState(state) {
        if (state.section) {
            this.section = state.section
        }
    }

    /**
     * Инициализирует ползователя при запуске приложения
     *
     * @returns {Promise<void>}
     */
    @once
    async authenticate() {
        try {
            let userAsPlain = (await Kopnik.fetchApi('get?ids='))[0]
            this.user = await Kopnik.get(userAsPlain.id)
            this.logger.info('user authenticated', this.user)
        } catch (err) {
            if ((err instanceof KopnikApiError) && err.message.match(/no.+aut/i)) {
                this.user = null
                this.logger.info('user not authenticated')
            } else {
                throw err
            }
        }
    }

    /**
     * Учстанавливает пользователем объект из VK.Auth callback
     *
     * @param vkUser
     * @returns
     */
    /*    async setVkUser(vkUser) {
            global.credentials = {
                uid: vkUser.uid,
                hash: vkUser.hash
            }
            this.user = await Kopnik.getByUid(vkUser.uid)
            if (!this.user) {
                this.user = new Kopnik
                this.user.merge({
                    firstName: vkUser.first_name,
                    lastName: vkUser.last_name,
                    photo: vkUser.photo,
                    smallPhoto: vkUser.photo_rec,
                })

                this.section = "Profile"
            }
            // this.user.uid= vkUser.uid
            // this.user.hash= vkUser.hash

            // localStorage.setItem("vkUser", JSON.stringify(vkUser))
        }
        */
}

Application.Section = {
    Main: 'Main',
    Profile: 'Profile',
    Witness: 'Witness',
    Thanks: 'Thanks',
}
