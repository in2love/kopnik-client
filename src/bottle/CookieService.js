/**
 * Предоставляет куки
 * По умолчанию куки из конфига config.di.coockie
 * Умеет пуш-попить куки, отходя от конфига при этом, полезно во время сквозных тестов
 * @example coockieService.push(1)
 * @example coockieService.pop()
 */
export default class CookieService {
    constructor(config) {
        this.stack=[null]
    }
    push(){
        return this.stack.push(this.cookie)
    }
    pop(){
        return this.stack.pop()
    }
    get cookie (){
        return this.stack[this.stack.length-1]
    }
    set cookie (value){
        this.stack[this.stack.length-1]=value
    }
}
