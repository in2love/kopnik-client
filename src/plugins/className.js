import * as models from "../models";

global.className=  function (constructor)
{         // uglify минимизирует имена классов, гаденыш https://github.com/vuejs/vue-cli/issues/1118
    if (constructor === models.Kopnik) {
        return 'Kopnik'
    }
    else{
        return constructor.name
    }
}
