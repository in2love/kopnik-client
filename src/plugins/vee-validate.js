import { required, email, max , min, size, numeric, length, digits, } from "vee-validate/dist/rules";
import { extend } from "vee-validate";

import i18n from "./i18n"

extend("required", {
    ...required,
});

extend("max", {
    ...max,
});

extend("email", {
    ...email,
});

extend("length", {
    ...length,
});
extend("numeric", {
    ...numeric,
});

import { localize } from 'vee-validate';
import en from 'vee-validate/dist/locale/en.json';
import ru from 'vee-validate/dist/locale/ru.json';

// Install English and Arabic locales.
localize({
    en,
    ru
})

// названия полей берутся из vue-i18n секция "profile"
let names= {}
for (let [eachLocale, eachLocaleMessages] of Object.entries(i18n.messages)){
    names[eachLocale]= {names: eachLocaleMessages.profile}
}
localize(names)

localize('ru');
