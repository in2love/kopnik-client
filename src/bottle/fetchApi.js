import {container} from '../plugins/bottle'
import _ from 'lodash'
import jsonToFormData from 'json-form-data'
import {KopnikApiError} from "../KopnikError";


export default async function fetchApi(url, options = {}) {
    const defaultOptions = {
            credentials: 'include',
            headers: {
                Accept: 'application/json',
                'Content-Type': !options.method || options.method.toUpperCase() == 'GET' ? 'text/plain' : 'application/x-www-form-urlencoded;charset=UTF-8',
            }
        },
        cookieOptions = container.cookieService.cookie ? {headers: {Cookie: container.cookieService.cookie}} : {}

    options = _.merge({}, defaultOptions, cookieOptions, options)
    if (options.body && options.headers['Content-Type'] === 'application/x-www-form-urlencoded;charset=UTF-8') {
        options.body = jsonToFormData(options.body)
    }
    // container.logger.warn(options)
    // console.log(container.config.api.path)
    let fullUrl = `${container.config.api.path}/${url}`.replace(/\w+\/\.\.\//, '')

    try {
        var response = await fetch(fullUrl, options)
        // Пропал 4G
    } catch (err) {
        throw new KopnikApiError(err.message, null, fullUrl)
    }
    // Не найдена страница или синтаксическая ошибка в веб-сервисе
    if (!response.ok) {
        throw new KopnikApiError(response.statusText, response.status, fullUrl)
    }

    let cookie = response.headers.get('set-cookie')
    if (cookie) {
        cookie = cookie.match(/(\w+=(\w|\d)+)/)[0]
        container.cookieService.cookie = cookie
    }
    // container.logger.warn(cookie, response.headers)
    let result
    switch (response.headers.get('Content-Type')) {
        case 'application/json':
            // result = await response.text()
            result = await response.json()
            // Не авторизован/Нет такого пользователя
            if (result.error) {
                throw new KopnikApiError(result.error.error_msg, result.error.error_code, fullUrl)
            }
            break
        default:
            result = await response.text()
    }

    return result.response
}
