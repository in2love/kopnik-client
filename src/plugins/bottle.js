import Bottle from "bottlejs"
import _ from "lodash"
import 'json-form-data'
import * as loglevel from "loglevel";

import Application from "../Application"
import config from '../../config'
import fetchApi from "../bottle/fetchApi";
// import fetchApiMock from "../bottle/fetchApi.mock";
import CookieService from "../bottle/CookieService"

Bottle.config.strict = true
const bottle = new Bottle()

bottle.service('cookieService', CookieService, 'config')
bottle.factory('fetchApi', function fetchApiFactory(container) {
    return container.config.di.fetch ? fetchApi : global.fetchApiMock
    // return fetchApi
})
bottle.factory('config', function configFactory() {
    if (!process.env.NODE_ENV) {
        throw new Error("NODE_ENV is not defined");
    }

    let local = {}//require("./local.js")
    let result = _.merge({}, config, local)[process.env.NODE_ENV]

    return result
})
if (!Application) {
    throw new Error("Application is undefined in bottle")
}
bottle.service('application', Application, 'logger')
bottle.factory('logger', function loggerFactory() {
    //все плагины ломают стектрейс консоли. то есть невозможно увидеть из какого файла и какой строки был вызван лог!
    // LoglevelPluginPrefix.reg(loglevel)
    // LoglevelPluginPrefix.apply(loglevel, {template: "%t [%l] %n: "})
    loglevel.setLevel(loglevel.levels.DEBUG)
    if (process.env.NODE_ENV == "test") {
        loglevel.setLevel(loglevel.levels.WARN)
        // LoglevelPluginToString.apply(loglevel, {})
    }
// Be sure to call setLevel method in order to apply plugin
// loglevel.getLogger("StateManager").setLevel("info")
    return loglevel
})

/**
 * @callback fetch
 * @param {string} url
 * @param {Object} options
 */
/**
 * @type {Object}
 *
 * @property {loglevel} logger
 * @property {Application} application
 * @property {Location} Location
 * @property {fetch} fetchApi
 * @property {CookieService} cookieService
 * @property {Object} config
 * @property {Object} config.api
 * @property {String} config.api.path
 * @property {Object} config.di
 * @property {Boolean} config.di.fetch
 * @property {Number} config.di.coockie
 * @property {Number} config.messenger.clientId
 * @property {String} config.messenger.loginUrl
 * @property {String} config.messenger.redirectUrl
 */
const container = bottle.container
export {bottle, container}
