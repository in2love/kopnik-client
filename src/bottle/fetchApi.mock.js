import {bottle, container} from '../plugins/bottle'
import {utils} from 'jest-snapshot'
import {resolve} from 'path'
import JSON5 from 'json5'
import {KopnikApiError} from "../KopnikError";

export default async function fetchApiMock(url, options = {}) {
    const snapshotPath = resolve(__dirname, '../../tests/system/__snapshots__/fetchApi.spec.js.snap'),
        snapshotData = utils.getSnapshotData(snapshotPath, false),
        user= container.cookieService.cookie||'anonymous',
        key = `system fetchApi ${user}@${url} 1`
    // console.log(snapshotPath, key, snapshotData)

    if (!container.cookieService.cookie && !url.match(/test.login.\d+/)) {
        throw new KopnikApiError('Not Authorized', 401, container.config.api.path + url)
    }

    if (snapshotData.data[key] === undefined) {
        throw new Error(`fetchApiMock could not find ${snapshotPath}[${key}]. use npm run test:system:watch -t fetchApi`)
    }
    const temp = snapshotData.data[key].replace(/Object |Array /g, '')
    // console.log(temp)
    if (url.match(/test.login.\d+/)) {
        const id = url.match(/test.login.(\d+)/)[1]
        container.cookieService.cookie = `user${id}`
    }
    return JSON5.parse(temp)
}
