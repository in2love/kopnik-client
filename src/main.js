import Vue from 'vue'
import vuetify from './plugins/vuetify'
import router from './plugins/vue-router'
import i18n from './plugins/i18n'
import "./plugins/vee-validate"

import {Icon} from 'leaflet'
import 'leaflet/dist/leaflet.css'
// this part resolve an issue where the markers would not appear
delete Icon.Default.prototype._getIconUrl;
Icon.Default.mergeOptions({
    iconRetinaUrl: require('leaflet/dist/images/marker-icon-2x.png'),
    iconUrl: require('leaflet/dist/images/marker-icon.png'),
    shadowUrl: require('leaflet/dist/images/marker-shadow.png')
})

import AppVue from './components/AppVue.vue'
import {container} from "./plugins/bottle"
import './plugins/className'

Vue.config.productionTip = false

global.application = container.application
global.application.authenticate()
/*    .then(() => {
        if (!this.user) {
            this.section = 'Map'
        } else if (this.user.status == Kopnik.Status.NEW || this.user.status == Kopnik.Status.DECLINED) {
            this.section = 'Profile'
        } else {
            this.section = 'Map'
        }
    })*/

global.vue = new Vue({
    ...AppVue,
    el: "#appContainer",
    vuetify,
    i18n,
    router,
    // render: h => h(App)
})//.$mount('#app')
