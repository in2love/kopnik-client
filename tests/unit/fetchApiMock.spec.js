import {KopnikApiError} from "../../src/KopnikError";

require('isomorphic-fetch');

import fetchApiMock from "../../src/bottle/fetchApi.mock";
import {bottle,container} from "../../src/plugins/bottle";



describe('unit.fetchApiMock', () => {
    beforeEach(() => {
        bottle.resetProviders(['cookieService'])
    })

    it('anonymous@test/login/1', async () => {
        let result = await fetchApiMock('test/login/1')
        expect(container.cookieService.cookie).toBe('user1')
        expect(result).toMatchSnapshot()
    });
    it('user1@users/get?ids=', async () => {
        await login(1)
        let result = await fetchApiMock('users/get?ids=')
        expect(result).toMatchSnapshot()
    });
    it('user1@users/get?ids=2', async () => {
        await login(1)
        let result = await fetchApiMock('users/get?ids=2')
        expect(result).toMatchSnapshot()
    });
    it('user1@users/get?ids=1,2,3', async () => {
        await login(1)
        let result = await fetchApiMock('users/get?ids=1,2,3')
        expect(result).toMatchSnapshot()
    })
})
