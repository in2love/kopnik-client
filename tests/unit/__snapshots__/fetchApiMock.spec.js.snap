// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit.fetchApiMock anonymous@test/login/1 1`] = `
Object {
  "message": "test login success",
  "status": "success",
  "user": Object {
    "birthyear": 1985,
    "firstName": "Павел",
    "foreman_id": null,
    "id": 1,
    "lastName": "Сысоев",
    "location": Array [
      47.042697,
      73.97209,
    ],
    "passport": 9991,
    "patronymic": "Якушев",
    "photo": "@todo",
    "smallPhoto": "@todo",
    "status": 2,
    "vk_id": 3809,
    "witness_id": null,
  },
}
`;

exports[`unit.fetchApiMock user1@users/get?ids= 1`] = `
Array [
  Object {
    "birthyear": 1985,
    "firstName": "Павел",
    "foreman_id": null,
    "id": 1,
    "lastName": "Сысоев",
    "location": Array [
      47.042697,
      73.97209,
    ],
    "passport": 9991,
    "patronymic": "Якушев",
    "photo": "@todo",
    "smallPhoto": "@todo",
    "status": 2,
    "witness_id": null,
  },
]
`;

exports[`unit.fetchApiMock user1@users/get?ids=1,2,3 1`] = `
Array [
  Object {
    "birthyear": 1985,
    "firstName": "Павел",
    "foreman_id": null,
    "id": 1,
    "lastName": "Сысоев",
    "location": Array [
      47.042697,
      73.97209,
    ],
    "passport": 9991,
    "patronymic": "Якушев",
    "photo": "@todo",
    "smallPhoto": "@todo",
    "status": 2,
    "witness_id": null,
  },
  Object {
    "birthyear": 1993,
    "firstName": "Максим",
    "foreman_id": null,
    "id": 2,
    "lastName": "Калашниковa",
    "location": Array [
      0,
      0,
    ],
    "passport": null,
    "patronymic": "Колесников",
    "photo": "@todo",
    "smallPhoto": "@todo",
    "status": 0,
    "witness_id": null,
  },
  Object {
    "birthyear": 1966,
    "firstName": "Валериан",
    "foreman_id": null,
    "id": 3,
    "lastName": "Некрасов",
    "location": Array [
      0,
      0,
    ],
    "passport": null,
    "patronymic": "Дементьев",
    "photo": "@todo",
    "smallPhoto": "@todo",
    "status": 0,
    "witness_id": null,
  },
]
`;

exports[`unit.fetchApiMock user1@users/get?ids=2 1`] = `
Array [
  Object {
    "birthyear": 1993,
    "firstName": "Максим",
    "foreman_id": null,
    "id": 2,
    "lastName": "Калашниковa",
    "location": Array [
      0,
      0,
    ],
    "passport": null,
    "patronymic": "Колесников",
    "photo": "@todo",
    "smallPhoto": "@todo",
    "status": 0,
    "witness_id": null,
  },
]
`;

exports[`unit.fetchApiMock users/get?ids= 1`] = `
Array [
  Object {
    "birthyear": 1231,
    "firstName": "Алексей",
    "foreman_id": null,
    "id": 2,
    "lastName": "Баранов",
    "location": Array [
      55.755826,
      37.6173,
    ],
    "passport": 1231,
    "patronymic": "qwer",
    "photo": "@todo",
    "smallPhoto": "@todo",
    "status": 2,
    "witness_id": null,
  },
]
`;

exports[`unit.fetchApiMock users/get?ids=1,2,3 1`] = `
Array [
  Object {
    "birthyear": 1234,
    "firstName": "Арт",
    "foreman_id": null,
    "id": 1,
    "lastName": "Арт",
    "location": Array [
      55.030494139179,
      82.918603420258,
    ],
    "passport": null,
    "patronymic": "Арт",
    "photo": "@todo",
    "smallPhoto": "@todo",
    "status": 2,
    "witness_id": 2,
  },
  Object {
    "birthyear": 1231,
    "firstName": "Алексей",
    "foreman_id": null,
    "id": 2,
    "lastName": "Баранов",
    "location": Array [
      55.755826,
      37.6173,
    ],
    "passport": 1231,
    "patronymic": "qwer",
    "photo": "@todo",
    "smallPhoto": "@todo",
    "status": 2,
    "witness_id": null,
  },
  Object {
    "birthyear": 1982,
    "firstName": "Александр",
    "foreman_id": null,
    "id": 3,
    "lastName": "Титов",
    "location": Array [
      55.158396343962,
      49.176560640553,
    ],
    "passport": null,
    "patronymic": "Николаевич",
    "photo": "@todo",
    "smallPhoto": "@todo",
    "status": 2,
    "witness_id": 2,
  },
]
`;

exports[`unit.fetchApiMock users/get?ids=2 1`] = `
Array [
  Object {
    "birthyear": 1231,
    "firstName": "Алексей",
    "foreman_id": null,
    "id": 2,
    "lastName": "Баранов",
    "location": Array [
      55.755826,
      37.6173,
    ],
    "passport": 1231,
    "patronymic": "qwer",
    "photo": "@todo",
    "smallPhoto": "@todo",
    "status": 2,
    "witness_id": null,
  },
]
`;
